src/js/actions/export.js

```diff
@@ -0,0 +1,222 @@
+     */
+    var _exportService;
+
+    /**
+     * before start up
+     * @return {Promise}
+     */
+    var beforeStartup = function () {
+        _exportService = new ExportService();
+        return _exportService.init();
+    };
+    beforeStartup.reads = [];
+    beforeStartup.writes = [];
+
+    var openExportPanel = function () {
+        return this.transfer(dialog.openDialog, "exports-panel-dialog");
+    };
+    openExportPanel.reads = [];
+    openExportPanel.writes = [locks.JS_DIALOG];
+
+
+    var _exportAsset = function (documentID, layer, asset) {
+        return _exportService.sendMessage("exportLayer", { layer: layer, scale: asset.scale })
+            .bind(this)
+            .then(function (exportResponse) {
+                // TODO we should inspect the response a little more closely, for export errors
+                if (exportResponse && Array.isArray(exportResponse.result)) {
+                    var assetProps = asset.toJS(),
+                        layerExportsMap = new Map(),
+                        payload;
+
+                    assetProps.filePath = exportResponse.result[0];
+                    assetProps.status = "stable"; // FIXME need a better way to handle these constants
+                    layerExportsMap.set(layer.id, assetProps);
+
+                    payload = {
+                        documentID: documentID,
+                        documentExports: {
+                            layerExportsMap: layerExportsMap
+                        }
+                    };
+                    this.dispatch(events.export.ASSET_CHANGED, payload);
+
+                    var documentExports = this.flux.stores.export.getDocumentExports(documentID),
+                        layerExportsArray = documentExports && documentExports.layerExportsArray(layer.id);
+
+                    return _updateLayerExportMetadata(documentID, layer.id, layerExportsArray);
+                } else {
+                    return Promise.resolve();
+                }
+            });
+    };
+
+    var exportAllAssets = function () {
+        var document = this.flux.stores.application.getCurrentDocument();
+
+        if (!document) {
+            Promise.resolve("No Document");
+        }
+
+        var documentID = document.id,
+            documentExports = this.flux.stores.export.getDocumentExports(documentID),
+            layerExportsMap = documentExports && documentExports.layerExportsMap,
+            exportArray = [];
+
+        if (!layerExportsMap || layerExportsMap.size < 1) {
+            return Promise.resolve("no assets to export");
+        }
+
+        layerExportsMap.forEach(function (layerExportAssets, layerID) {
+            var layer = document.layers.byID(layerID);
+            layerExportAssets.forEach(function (asset) {
+                exportArray.push({ layer: layer, asset: asset });
+            }, this);
+        }, this);
+
+        // TODO this is intentionally serial for now, because or WS is lame 
```

Comment:
Just to be clear, this will *no longer be serial* now that we're using NodeDomain.  My original lame websocket implementation was braindead and single threaded.